### PR TITLE
added bilinear() to tf in control lti

### LINKED ISSUE #24544: bilinear discretization of continuous transfer function in Control API
Continuous transfer functions $$H(s)=\frac{b_0+b_1 s+b_2 s^2+...}{a_0+b_1 s+a_2 s^2+...}$$

can be discretized to $H(z)$ with the bilinear (Möbius/Tustin) formula (leads to trapezoidal integration rule in the time domain)

$$s(z)=\frac{2}{T}\frac{z-1}{z+1}$$

where $T$ is the sample period, $H(z)$ is important when implementing digital control strategies.

The expression s(z) is substitued into $H(s)$ and $H(z)$ is obtained, see

[https://en.wikipedia.org/wiki/Bilinear_transform](https://en.wikipedia.org/wiki/Bilinear_transform)

A similar Möbius or bilinear transform is used in a totaly different engineering application (fluid power engineering) in

Lauer-Baré Z. and Gaertig E., [*Conformal Mappings with SymPy: Towards Python-driven Analytical Modeling in Physics*. Lauer-Baré, Z. & Gaertig, E. In Agarwal, M., Calloway, C., Niederhut, D., & Shupe, D., editors, Proceedings of the 20th Python in Science Conference, pages 85 - 93, 2021. ](https://conference.scipy.org/proceedings/scipy2021/lauer_bare_gaertig.html)

The conference talk can be seen on the [Enthought YouTube channel](https://www.youtube.com/watch?v=P5ybpjv2uDA).

References on the Möbius/bilinear transform in the context of discretizations of continuous transfer functions can be found here

Oppenheim AV. Discrete-time signal processing. Pearson Education; 1999.

Lyons RG. Understanding digital signal processing, 3/E. Pearson Education; 1997.

The following bilinear function for SymPy's Control API

https://docs.sympy.org/latest/modules/physics/control/lti.html

exists already in numerical form in 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.bilinear.html

Here is the proposed bilinear transform,

    """
    Discretizes H(s) (just a Sympy rational polynomial in) to H(z) with the bilinear Method and ample rate T
    It Returns falling coeffs vof H(z) from numerator and denominator
    i.e. H(z) = (az+b)/(cz+d) is return as [a, b], [c, d]
    
    The latex formulae of the continuous input formula and the dicretized numerator and denominator are returned
    
    """

  """

def bilinear(H):
   
    print('H(s)='+sym.latex(H))
    print()
    
    HZ = H.subs(s, (2/T)*(z-1)/(z+1))

    zaehler, nenner = sym.fraction(sym.simplify(HZ))
    
    HZzaehler = sym.collect(sym.expand(zaehler, z), z)
    
    HZzaehler = sym.poly(HZzaehler, z)
    
    HZnenner = sym.collect(sym.expand(nenner, z), z)
    
    HZnenner = sym.poly(HZnenner, z)
    
    print('discrete numerator:\n' + sym.latex(HZzaehler.expr))
    print()
    
    print('discrete denominator:\n' +sym.latex(HZnenner.expr))    
    
    zaehler_coefs = HZzaehler.coeffs()
    
    nenner_coefs = HZnenner.coeffs()
    
    return zaehler_coefs, nenner_coefs

  """
A jupyter notebook example is attached.

[sympy_contrlol_api_discretize_bilinear.zip](https://github.com/sympy/sympy/files/10445059/sympy_contrlol_api_discretize_bilinear.zip)

### PR DESCRIPTION
**Pull request by zolabar**
# bilinear discretization of continuous transfer function in Control API

#### References to other Issues or PRs

Fixes #24544

#### Brief description of what is fixed or changed

Added the function bilinear() to TransferFunction objects in sympy\physics\control\lti.py

It returns falling coeffs of H(z) from numerator and denominator. H(z) is the corresponding discretized tf, discretized with bilinear transform method. Coefficients are falling, i.e. H(z) = (az+b)/(cz+d) is returned as [a, b], [c, d].

#### Other comments

It is a useful function for numerical implementations of discrete signal processing and control with scipy.signal, control or matlab/simulink, once the symbolic continuous transfer function design is finished.

#### Release Notes

* functions
  * Added `bilinear`, a function for the bilinear transform discretization of a continuous transfer function from Control API.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
> The following commits **add new files**:
> 
> * [fe659ef](https://github.com/sympy/sympy/commit/fe659effc221103a8cfab9d13af59c5aa6bde071):
>   
>   * `examples/notebooks/transfer_function_bilinear.ipynb`
>   * `transfer_function_bilinear.ipynb`
> 
> The following commits **delete files**:
> 
> * [727fd6b](https://github.com/sympy/sympy/commit/727fd6b6a0166882a266e1aa34996da2932c2071):
>   
>   * `transfer_function_bilinear.ipynb`

These commits will need to be squashed. We can't have a commit that introduces a file followed by another that deletes it. The file must not exist in the history at all.

File: sympy/physics/control/lti.py [hunk 1]
```diff
@@ -679,6 +683,52 @@ def __pow__(self, p):
...
+        [T, T]
+        >>> denZ
+        [2*L + R*T, -2*L + R*T]
+        """        
+        
+        z = Symbol('z') # discrete variable z 
+        T = Symbol('T')  # and sample period T
+       
+        H = self.num/self.den
+
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:714]: I see lot of blank lines in between the code which are not required.

File: sympy/physics/control/lti.py [hunk 2]
```diff
@@ -679,6 +683,52 @@ def __pow__(self, p):
...
+        >>> numZ
+        [T, T]
+        >>> denZ
+        [2*L + R*T, -2*L + R*T]
+        """        
+        
+        z = Symbol('z') # discrete variable z 
+        T = Symbol('T')  # and sample period T
+       
+        H = self.num/self.den
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:713]: One workaround I see here which I would consider fundamentally incorrect is taking input in the form of a `TransferFunction` but then converting it into a symbolic expression and then working on it.

```
>>> from sympy import *
>>> from sympy.physics.control.lti import TransferFunction
>>> from sympy.abc import s, L, R, T, z
>>> tf = TransferFunction(1, s*L + R, s)
>>> type(tf)
<class 'sympy.physics.control.lti.TransferFunction'>
>>> H = tf.num/tf.den
>>> type(H)
<class 'sympy.core.power.Pow'>
```

This is not incorrect, but we are aware that `subs`, `collect`, `expand` do not work with the 'TransferFunction' class as of now. I would prefer that to be modified and then the entire function to be working on a variable of type - `<class 'sympy.physics.control.lti.TransferFunction'>`
Basically, if we are changing something then let's to it fully right. But if others agree then we can proceed with this too (it is clearly a useful addition).

**Comment by faze-geek** [comment]
As far as the failing test is concerned you should run` python bin/strip_whitespace sympy/physics/control/lti.py`. Then your code quality check will pass.

**Comment by 1e9abhi1e10** [comment]
you need to add your name and email in the `.mailmap` file and then run `python bin/mailmap_check.py` for instructions read [here](https://github.com/sympy/sympy/wiki/Development-workflow#add-your-name-and-email-address-to-the-mailmap-file).

**Comment by zolabar** [comment]
> you need to add your name and email in the `.mailmap` file and then run `python bin/mailmap_check.py` for instructions read [here](https://github.com/sympy/sympy/wiki/Development-workflow#add-your-name-and-email-address-to-the-mailmap-file).

thanks, just did that

**Comment by zolabar** [comment]
Check the .mailmap file

**Comment by 1e9abhi1e10** [comment]
> Check the .mailmap file
> […](#)
> Am 21.01.2023 um 08:51 schrieb ABHISHEK PATIDAR ***@***.***>: ﻿ you need to add your name and email in the .mailmap file and then run python bin/mailmap_check.py for instructions read here. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.

After adding your name and email in the `.mailmap file` you need to run `python bin/mailmap_check.py` and then push a new commit

**Comment by zolabar** [comment]
> As far as the failing test is concerned you should run` python bin/strip_whitespace sympy/physics/control/lti.py`. Then your code quality check will pass.



> As far as the failing test is concerned you should run` python bin/strip_whitespace sympy/physics/control/lti.py`. Then your code quality check will pass.

Thanks, now all checks have passed :)

File: sympy/physics/control/lti.py [hunk 3]
```diff
@@ -302,6 +303,9 @@ def __new__(cls, num, den, var):
         else:
             raise TypeError("Unsupported type for numerator or denominator of TransferFunction.")
 
+
+
+
```
Comment (sylee957) [inline sympy/physics/control/lti.py:308]:
```suggestion
```

Comment (sylee957) [inline sympy/physics/control/lti.py:308]:
```suggestion
```

**Comment by faze-geek** [comment]
> Thanks, now all checks have passed :)

Could you add relevant examples in form of tests in https://github.com/sympy/sympy/blob/master/sympy/physics/control/tests/test_lti.py

**Comment by zolabar** [comment]
> > Thanks, now all checks have passed :)
> 
> Could you add relevant examples in form of tests in https://github.com/sympy/sympy/blob/master/sympy/physics/control/tests/test_lti.py

Ok, i‘ll prepare some tests

**Comment by zolabar** [comment]
> Could you add relevant examples in form of tests in https://github.com/sympy/sympy/blob/master/sympy/physics/control/tests/test_lti.py

@faze-geek,  done [e595caf](https://github.com/sympy/sympy/pull/24558/commits/e595caf477e53204489300ee219e7defcb6a3c6c)

**Comment by faze-geek** [comment]
> @faze-geek, done [e595caf](https://github.com/sympy/sympy/pull/24558/commits/e595caf477e53204489300ee219e7defcb6a3c6c)

Thanks, I would appreciate if any maintainer of `control` module could review it. @akshanshbhatt Do you mind taking a look at this !?

**Comment by hanspi42** [comment]
I think two things should be changed:
1. The method should take `z` and `T` as input arguments.
2. This can be done without creating a symbolic expression internally, as @faze-geek says: if `N` is the maximum of the degrees of numerator and denominator, then multiplying numerator and denominator by `(z+1)**N` after substitution gives polynomials in `z`.  E.g., if the numerator polynnomial is `a*s**2+b*s+c`, and `N=3` because we have a third-order denominator, the new numerator polynomial is `a*(2/T)**2*(z-1)**2*(z+1) + b*(2/T)*(z-1)*(z+1)**2 + c*(z+1)**3`. This looks simple to code.

**Comment by zolabar** [comment]
> I think two things should be changed:
> 
> 1. The method should take `z` and `T` as input arguments.
> 2. This can be done without creating a symbolic expression internally, as @faze-geek says: if `N` is the maximum of the degrees of numerator and denominator, then multiplying numerator and denominator by `(z+1)**N` after substitution gives polynomials in `z`.  E.g., if the numerator polynnomial is `a*s**2+b*s+c`, and `N=3` because we have a third-order denominator, the new numerator polynomial is `a*(2/T)**2*(z-1)**2*(z+1) + b*(2/T)*(z-1)*(z+1)**2 + c*(z+1)**3`. This looks simple to code.

@hanspi42, thank you for the feedback.

Do I understand correctly, that  concerning point 1, you propose that the function should be called as (for tf as TransferFunction, discrete_var the symbolic discrete variable (e.g. z) and sample_per the symbolic sample period) :

`
tf.blinear(discrete_var, sample_per)
`
?

Concerning point 2, I don't see how to take advantage of your analysis, that is, how should I get the list of the numerator and denominator coeffcients of the discretized transfer function?

Taking into account your point 1. it could look as follows. However, I as I wrote above, I have no idea of how to avoid the defining ` H = self.num/self.den` and then using SymPy to get the list of the numerator and denominator coeffcients of the discretized transfer function (which are needed for discrete implementation in practice then)...


`def tf.blinear(self, discrete_var, sample_per):


        z = discrete_var
        T = sample_per
      
        H = self.num/self.den
        HZ = H.subs(self.var, (2/T)*(z-1)/(z+1))

        num, den = fraction(HZ.simplify())

        HZnum = collect(expand(num, z), z)
        HZnum = Poly(HZnum, z)

        HZden = collect(expand(den, z), z)
        HZden = Poly(HZden, z)

        num_coefs = HZnum.coeffs()
        den_coefs = HZden.coeffs()

        return num_coefs, den_coefs

`

**Comment by hanspi42** [comment]
> Concerning point 2, I don't see how to take advantage of your analysis, that is, how should I get the list of the numerator and denominator coeffcients of the discretized transfer function?

For example: 
```
from sympy import *
s, z = symbols('s z')
T, a, b, c, d = symbols('T a b c d', real=True)

n = a*s**2+b*s+c
d = b*s**3+c*s**2+d*s+a
H = n/d
display(H)

np = n.as_poly(s).all_coeffs()
dp = d.as_poly(s).all_coeffs()

# This is the suggested code starting from having numerator and denominator
# coefficients np and dp
N = max(len(np), len(dp)) - 1
n1 = Add(*[ T**(N-i)*2**i*c*(z-1)**i*(z+1)**(N-i) for c, i in zip(np[::-1], range(len(np))) ])
d1 = Add(*[ T**(N-i)*2**i*c*(z-1)**i*(z+1)**(N-i) for c, i in zip(dp[::-1], range(len(dp))) ])

# This shows the transfer function
H1 = n1/d1
display(H1)

# This is the solution you are using
H2 = H.subs(s, 2/T*(z-1)/(z+1))
display(H2)

# This shows that it is the same; must be zero.
dH = (H1-H2).simplify()
display(dH)
```

**Comment by hanspi42** [comment]
> tf.blinear(discrete_var, sample_per)

Yes ;)

**Comment by zolabar** [comment]
> > Concerning point 2, I don't see how to take advantage of your analysis, that is, how should I get the list of the numerator and denominator coeffcients of the discretized transfer function?
> 
> For example:
> 
> ```
> from sympy import *
> s, z = symbols('s z')
> T, a, b, c, d = symbols('T a b c d', real=True)
> 
> n = a*s**2+b*s+c
> d = b*s**3+c*s**2+d*s+a
> H = n/d
> display(H)
> 
> np = n.as_poly(s).all_coeffs()
> dp = d.as_poly(s).all_coeffs()
> 
> # This is the suggested code starting from having numerator and denominator
> # coefficients np and dp
> N = max(len(np), len(dp)) - 1
> n1 = Add(*[ T**(N-i)*2**i*c*(z-1)**i*(z+1)**(N-i) for c, i in zip(np[::-1], range(len(np))) ])
> d1 = Add(*[ T**(N-i)*2**i*c*(z-1)**i*(z+1)**(N-i) for c, i in zip(dp[::-1], range(len(dp))) ])
> 
> # This shows the transfer function
> H1 = n1/d1
> display(H1)
> 
> # This is the solution you are using
> H2 = H.subs(s, 2/T*(z-1)/(z+1))
> display(H2)
> 
> # This shows that it is the same; must be zero.
> dH = (H1-H2).simplify()
> display(dH)
> ```

@hanspi42 , thank you for the fast reply and the elegant compact code example, i‘ll test it and think about it!

**Comment by zolabar** [comment]
> > tf.blinear(discrete_var, sample_per)
> 
> Yes ;)

@hanspi42 , commited the changes, thank you very much for the algebraic hints and the code prototype.
Since the discrete variable does not appear in the output coeff list I only included the sample period as input, hence the current function is used as

`
nz, dz = tf.bilinear(T)
`

**Comment by sylee957** [comment]
I think that general interface is good. 
But I'm not sure it should be added under methods because transfer function basic classes only supports arithmetic now

I think that we can pull out as function outside if we are not clear.

**Comment by sylee957** [comment]
I think that adding additional features are okay if it is outside, but only want to avoid adding inside objects as much as possible because of [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle)

**Comment by zolabar** [comment]
> I think that general interface is good. But I'm not sure it should be added under methods because transfer function basic classes only supports arithmetic now
> 
> I think that we can pull out as function outside if we are not clear.

@sylee957 , thank you for the feedback.

For me it would be OK to inlcude `bilinear` as an outside function, like in

[https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.bilinear.html](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.bilinear.html)

**Comment by zolabar** [comment]
@sylee957,

in commit

[https://github.com/sympy/sympy/pull/24558/commits/36d6d4a608b6a991e2dc5acc31af6b89e11aaf85(https://github.com/sympy/sympy/pull/24558/commits/36d6d4a608b6a991e2dc5acc31af6b89e11aaf85) 

`bilinear` is a function, used as

`bilinear(tf, T)`

similar to its numerical counterpart from sicipy.signal

File: sympy/physics/control/lti.py (hunk 2 above)
Comment (zolabar) [reply sympy/physics/control/lti.py]: @faze-geek , That proposal is realized in commit

[https://github.com/sympy/sympy/commit/36d6d4a608b6a991e2dc5acc31af6b89e11aaf85](https://github.com/sympy/sympy/commit/36d6d4a608b6a991e2dc5acc31af6b89e11aaf85)

thanks to comments, suggestions and contributions of @sylee957 and @hanspi42

Comment (faze-geek) [reply sympy/physics/control/lti.py]: Thanks @zolabar , looks good to me. You can mark this conversation resolved accordingly.

**Comment by hanspi42** [comment]
It is OK for me now, but I would suggest one final change: show the formula of the bilinear-transform substitution in the doc string.

**Comment by zolabar** [comment]
> It is OK for me now, but I would suggest one final change: show the formula of the bilinear-transform substitution in the doc string.

-> done

**Comment by hanspi42** [comment]
> > It is OK for me now, but I would suggest one final change: show the formula of the bilinear-transform substitution in the doc string.
> 
> -> done

It looks OK, but does not appear in the documentation (preview here: https://output.circle-artifacts.com/output/job/1eb985ff-0719-4e5a-a222-ce27d1c7ee08/artifacts/0/doc/_build/html/index.html)

Unfortunately I do not know how to do this; does any of the other reviewers know?

**Comment by sylee957** [comment]
You need to add a directive `..autofunction` to `doc/src/modules/physics/control/lti.rst`

**Comment by zolabar** [comment]
> You need to add a directive `.. autofunction: bilinear` to `doc/src/modules/physics/control/lti.rst`

-> done

**Comment by zolabar** [comment]
> You need to add a directive `.. autofunction: bilinear` to `doc/src/modules/physics/control/lti.rst`

-> done

**Comment by zolabar** [comment]
> It is OK for me now, but I would suggest one final change: show the formula of the bilinear-transform substitution in the doc string.

@hanspi42 , i included the docstring that @sylee957 suggested. How did you generate the test preview of the documentation? Thanks

**Comment by hanspi42** [comment]
> How did you generate the test preview of the documentation?

This is done by github. Look under "All checks have passed", there is an entry "Click here to see a preview of the documentation"

**Comment by hanspi42** [comment]
@sylee957: This now looks OK to me.

**Comment by sylee957** [comment]
I think that docs should be more readable because summary is too long and there are not many examples to figure out how it works for general cases.

**Comment by sylee957** [comment]
Also, I find that the release notes are misleading to `functions`, I'd correct them manually

File: .mailmap [hunk 4]
```diff
@@ -1526,6 +1527,7 @@ wookie184 <wookie1840@gmail.com>
 wuyudi <wuyudi119@163.com>
 ylemkimon <ylemkimon@naver.com> <mail@ylem.kim>
 znxftw <vishnu2101@gmail.com>
+zolabar <raszoufine@gmail.com>
```
Comment (oscarbenjamin) [inline .mailmap:1530]: You have added two separate entries to the .mailmap file:
https://github.com/sympy/sympy/blob/4fb6b42dd0941f05178aa570edf9ef61162ce960/.mailmap#L138-L148
I'm going to presume that you want your name and email to be recorded as:
```
Zoufiné Lauer-Baré <raszoufine@gmail.com>
```

Comment (oscarbenjamin) [reply .mailmap:1530]: I'm changing this in #24775. Please comment there if you would like a different name/email.

Comment (zolabar) [reply .mailmap:1530]: Thank you
